zcash_primitives/src/note_encryption.rs

```diff
@@ -335,52 +340,87 @@ impl SaplingNoteEncryption {
     }
 }
 
-fn parse_note_plaintext_without_memo(
+fn parse_note_plaintext_without_memo<P: consensus::Parameters>(
+    height: u32,
     ivk: &Fs,
+    epk: &edwards::Point<Bls12, PrimeOrder>,
     cmu: &Fr,
     plaintext: &[u8],
 ) -> Option<(Note<Bls12>, PaymentAddress<Bls12>)> {
     // Check note plaintext version
-    match plaintext[0] {
-        0x01 => (),
-        _ => return None,
+    if !plaintext_version_is_valid::<P>(height, plaintext[0]) {
+        return None;
     }
 
     let mut d = [0u8; 11];
     d.copy_from_slice(&plaintext[1..12]);
 
     let v = (&plaintext[12..20]).read_u64::<LittleEndian>().ok()?;
 
-    let rcm = Fs::from_repr(FsRepr(
-        plaintext[20..COMPACT_NOTE_SIZE]
-            .try_into()
-            .expect("slice is the correct length"),
-    ))?;
+    let r: [u8; 32] = plaintext[20..COMPACT_NOTE_SIZE]
+        .try_into()
+        .expect("slice is the correct length");
+
+    let rseed = if plaintext[0] == 0x01 {
+        let rcm = Fs::from_repr(FsRepr(r))?;
+        Rseed::BeforeZip212(rcm)
+    } else {
+        Rseed::AfterZip212(r)
+    };
 
     let diversifier = Diversifier(d);
     let pk_d = diversifier
         .g_d::<Bls12>(&JUBJUB)?
         .mul(ivk.to_repr(), &JUBJUB);
 
     let to = PaymentAddress::from_parts(diversifier, pk_d)?;
-    let note = to.create_note(v, rcm, &JUBJUB).unwrap();
+    let note = to.create_note(v, rseed, &JUBJUB).unwrap();
 
     if note.cm(&JUBJUB) != *cmu {
         // Published commitment doesn't match calculated commitment
         return None;
     }
 
+    if let Rseed::AfterZip212(_) = note.rseed {
+        let derived_esk = note.derive_esk().unwrap();
```

Comment:
```suggestion
    if let Some(derived_esk) = note.derive_esk() {
```

Reply:
This (and the other similar refactorings below) avoids an `unwrap` that could potentially panic if we got it wrong.